zkevm-circuits/src/evm_circuit/op_execution/jump.rs

```diff
@@ -0,0 +1,191 @@
+use super::super::{
+    Case, Cell, Constraint, CoreStateInstance, ExecutionStep, Word,
+};
+use super::utils::{
+    self,
+    common_cases::{OutOfGasCase, StackUnderflowCase},
+    constraint_builder::ConstraintBuilder,
+    StateTransition, StateTransitionExpressions,
+};
+use super::{CaseAllocation, CaseConfig, OpExecutionState, OpGadget};
+use crate::impl_op_gadget;
+use crate::util::{Expr, ToWord};
+use bus_mapping::evm::{GasCost, OpcodeId};
+use halo2::plonk::Error;
+use halo2::{arithmetic::FieldExt, circuit::Region};
+use num::ToPrimitive;
+use std::convert::TryInto;
+
+static STATE_TRANSITION: StateTransition = StateTransition {
+    gc_delta: Some(1),
+    pc_delta: Some(0), // TODO: pc is dynamic for jump
+    sp_delta: Some(1),
+    gas_delta: Some(GasCost::MID.as_u64()),
+    next_memory_size: None,
+};
+
+const NUM_POPPED: usize = 1;
+impl_op_gadget!(
+    #set[JUMP]
+    JumpGadget {
+        JumpSuccessCase(),
+        OutOfGasCase(STATE_TRANSITION.gas_delta.unwrap()),
+        StackUnderflowCase(NUM_POPPED),
+        //TODO: ErrJumpcase
+    }
+);
+
+#[derive(Clone, Debug)]
+struct JumpSuccessCase<F> {
+    case_selector: Cell<F>,
+    code_hash: Word<F>,
+    dest: Word<F>,
+}
+
+impl<F: FieldExt> JumpSuccessCase<F> {
+    pub(crate) const CASE_CONFIG: &'static CaseConfig = &CaseConfig {
+        case: Case::Success,
+        num_word: 2, // top stack value as pc, hash of contract
+        num_cell: 0,
+        will_halt: false,
+    };
+
+    pub(crate) fn construct(alloc: &mut CaseAllocation<F>) -> Self {
+        Self {
+            case_selector: alloc.selector.clone(),
+            code_hash: alloc.words.pop().unwrap(),
+            dest: alloc.words.pop().unwrap(),
+        }
+    }
+
+    pub(crate) fn constraint(
+        &self,
+        state_curr: &OpExecutionState<F>,
+        state_next: &OpExecutionState<F>,
+        name: &'static str,
+    ) -> Vec<Constraint<F>> {
+        let mut cb = ConstraintBuilder::default();
+
+        // State transitions
+        // `pc` needs to be `dest` value
+        let mut st = StateTransitionExpressions::new(STATE_TRANSITION.clone());
+        st.pc_delta =
+            Some(self.dest.expr() - state_curr.program_counter.expr());
```

Comment:
It should be recomposite bytes here instead of using the random linear combination, right? And we at least need to verify the MSB is `0` to make sure it doesn't wrap around the field. (or ideally we only need 3 bytes instead of 32 bytes since there are some bound of creation code or deployed code, we can further do this with the new constrain builder introduced in #196 later.)

```suggestion
            Some(from_bytes::expr(self.dest.cells) - state_curr.program_counter.expr());
```

Reply:
I think  it's right to  decomposite to extract the real 'dest' value as it maybe multi bytes length,  secondly,  in the successful case, lookup byte code table successfully can ensure 'dest' is valid without checking max 3 bytes ,  do you think so ?

Reply:
no, I mean 32 bytes recomposition would wrap around the field, which we need to check in evm circuit.

Reply:
I see,   currently I would prefer change to "Some(from_bytes::expr(self.dest.cells[..3]) - state_curr.program_counter.expr()); "  here,   the further moving on #196 later

Reply:
`from_bytes::expr(self.dest.cells[..3])` for now makes sense to me.

Reply:
fixed in 8882a72 thanks for catching it.